File: bindings/ruby/test/adapter_test.rb
# frozen_string_literal: true

require 'minitest/autorun'
require 'json'
require 'strling'

class AdapterTest < Minitest::Test
  def test_canonical_request_and_lexical_helper_shapes
    request = Strling.source_compile_request('literal "hello"')
    assert_equal '1.0.0', request['contract_version']
    assert_equal 'semantic_strling', request.dig('input', 'document', 'frontend', 'id')
    assert_equal 'stdlib.email', Strling.email('root').dig('arguments', 'helper_id')
    assert_nil Strling.uuid('root').dig('arguments', 'parameters', 'version')
  end

  def test_stdlib_helpers_consume_canonical_essential_fixture
    fixture_path = File.expand_path('../../../spec/stdlib/essential_5.json', __dir__)
    patterns = JSON.parse(File.read(fixture_path, encoding: 'UTF-8')).fetch('patterns')
    helpers = {
      'dateTime' => Strling.date_time('date-time'),
      'email' => Strling.email('email'),
      'ip' => Strling.ip('ip'),
      'url' => Strling.url('url'),
      'uuid' => Strling.uuid('uuid')
    }

    assert_equal helpers.keys.sort, patterns.keys.sort
    helpers.each do |name, step|
      helper_id = name == 'dateTime' ? 'date_time' : name
      assert_equal "stdlib.#{helper_id}", step.dig('arguments', 'helper_id')
    end
  end

  def test_relative_path_fails_closed
    error = assert_raises(Strling::NativeAdapterError) { Strling.load_native('libstrling_interop.so') }
    assert_equal :native_load, error.kind
    nul_error = assert_raises(Strling::NativeAdapterError) { Strling.load_native("/tmp/strling\0invalid") }
    assert_equal :native_load, nul_error.kind
  end

  def test_live_native_transport_when_certification_probe_is_supplied
    path = ENV['STRLING_DYNAMIC_PROBE']
    skip 'certification probe unavailable' if path.nil?

    client = Strling.load_native(path)
    expected = { 'unicode' => '雪' }
    results = {
      'describe' => client.describe,
      'compile' => client.compile(Strling.source_compile_request('literal "雪"')),
      'target_profile.inspect' => client.inspect_target_profile('contract_version' => '1.0.0'),
      'simply.compile' => client.simply_compile(Strling.simply_builder_request([Strling.email('root')], 'root'))
    }
    results.each_value { |result| assert_equal expected, result }
    evidence_dir = ENV['STRLING_DYNAMIC_EVIDENCE_DIR']
    File.write(File.join(evidence_dir, 'ruby.json'), JSON.generate(results)) unless evidence_dir.nil?
    concurrent = Strling.load_native(ENV.fetch('STRLING_DYNAMIC_CONCURRENCY_PROBE', path))
    workers = Array.new(4) do
      Thread.new do
        4.times do
          raise 'concurrent canonical result drifted' unless concurrent.describe == expected
        end
        true
      end
    end
    workers.each { |worker| assert worker.value }
    concurrent.close
    client.close
    assert client.closed?
    client.close
    assert_raises(Strling::NativeAdapterError) { client.describe }
  end

  def test_transport_probes_fail_closed
    {
      'STRLING_DYNAMIC_ABI_PROBE' => :native_abi,
      'STRLING_DYNAMIC_OVERSIZE_PROBE' => :transport,
      'STRLING_DYNAMIC_DUPLICATE_PROBE' => :transport,
      'STRLING_DYNAMIC_INVALID_UTF8_PROBE' => :transport,
      'STRLING_DYNAMIC_RELEASE_FAILURE_PROBE' => :native_abi
    }.each do |name, kind|
      path = ENV[name]
      next if path.nil?

      error = assert_raises(Strling::NativeAdapterError) do
        client = Strling.load_native(path)
        client.describe
      end
      assert_equal kind, error.kind
    end
  end
end

File: bindings/dart/test/adapter_test.dart
import 'dart:convert';
import 'dart:io';
import 'dart:isolate';

import 'package:path/path.dart' as path;
import 'package:strling/strling.dart';
import 'package:test/test.dart';

void main() {
  test('source request contains canonical envelope data', () {
    final request = sourceCompileRequest("'hello'");
    expect(request['contract_version'], '1.0.0');
    final input = request['input']! as Map<String, Object?>;
    final document = input['document']! as Map<String, Object?>;
    expect(document['source_id'], 'src:dart.adapter');
  });

  test('stdlib helper records identity without semantics', () {
    expect(email('root'), {
      'step_id': 'root',
      'operation': 'stdlib_helper',
      'arguments': {
        'helper_id': 'stdlib.email',
        'parameters': <String, Object?>{},
      },
    });
  });

  test('relative native path fails closed', () {
    expect(
      () => NativeClient.load('relative/strling'),
      throwsA(
        isA<NativeAdapterException>().having(
          (error) => error.kind,
          'kind',
          NativeErrorKind.load,
        ),
      ),
    );
  });

  test('native adapter preserves canonical operations and lifecycle', () async {
    final configured = Platform.environment['STRLING_NATIVE_LIBRARY'];
    if (configured == null || configured.isEmpty) {
      markTestSkipped(
        'STRLING_NATIVE_LIBRARY is required for governed integration execution',
      );
      return;
    }
    final nativePath = path.normalize(path.absolute(configured));
    final client = NativeClient.load(nativePath);
    final describe = client.describe();
    final compile = client.compile(
      sourceCompileRequest(
        'literal "héllo\u0000世界"',
        sourceId: 'src:adapter.parity',
      ),
    );
    final profileFile = File(
      path.normalize(
        path.join(
          Directory.current.path,
          '..',
          '..',
          'spec',
          'targets',
          'profiles',
          'pcre2-10.43.json',
        ),
      ),
    );
    final profile = jsonDecode(profileFile.readAsStringSync())
        as Map<String, Object?>;
    final inspected = client.inspectTargetProfile(profile);
    final builder = simplyBuilderRequest([email('parity-root')], 'parity-root')
      ..['identity_namespace'] = 'adapter-parity';
    final simply = client.simplyCompile(builder);

    final concurrent = await Future.wait(
      List.generate(
        8,
        (_) => Isolate.run(() {
          final isolated = NativeClient.load(nativePath);
          try {
            return isolated.describe() is Map<String, Object?>;
          } finally {
            isolated.close();
          }
        }),
      ),
    );
    expect(concurrent, everyElement(isTrue));

    _writeEvidence({
      'describe': describe,
      'compile': compile,
      'target_profile': inspected,
      'simply': simply,
    });
    client.close();
    client.close();
    expect(
      client.describe,
      throwsA(
        isA<NativeAdapterException>().having(
          (error) => error.kind,
          'kind',
          NativeErrorKind.closed,
        ),
      ),
    );
  }, timeout: const Timeout(Duration(minutes: 2)));

  test('native release probe requires same-descriptor free', () {
    final configured = Platform.environment['STRLING_GDS_RELEASE_PROBE'];
    if (configured == null || configured.isEmpty) {
      markTestSkipped(
        'STRLING_GDS_RELEASE_PROBE is required for governed lifecycle execution',
      );
      return;
    }
    final client = NativeClient.load(path.normalize(path.absolute(configured)));
    try {
      expect(client.describe(), isA<Map<String, Object?>>());
      expect(client.describe(), isA<Map<String, Object?>>());
    } finally {
      client.close();
    }
  });

  test('native ABI mismatch fails before execution', () {
    final configured = Platform.environment['STRLING_GDS_ABI_PROBE'];
    if (configured == null || configured.isEmpty) {
      markTestSkipped('STRLING_GDS_ABI_PROBE is required for governed ABI execution');
      return;
    }
    expect(
      () => NativeClient.load(path.normalize(path.absolute(configured))),
      throwsA(
        isA<NativeAdapterException>()
            .having((error) => error.kind, 'kind', NativeErrorKind.abi)
            .having((error) => error.status, 'status', 2),
      ),
    );
  });

  test('native oversized response fails closed', () {
    final configured = Platform.environment['STRLING_GDS_OVERSIZE_PROBE'];
    if (configured == null || configured.isEmpty) {
      markTestSkipped(
        'STRLING_GDS_OVERSIZE_PROBE is required for governed bounds execution',
      );
      return;
    }
    final client = NativeClient.load(path.normalize(path.absolute(configured)));
    try {
      expect(
        client.describe,
        throwsA(
          isA<NativeAdapterException>().having(
            (error) => error.kind,
            'kind',
            NativeErrorKind.transport,
          ),
        ),
      );
    } finally {
      client.close();
    }
  });

  test('native duplicate response fails closed', () {
    _expectTransportProbe('STRLING_GDS_DUPLICATE_PROBE');
  });

  test('native invalid UTF-8 response fails closed', () {
    _expectTransportProbe('STRLING_GDS_INVALID_UTF8_PROBE');
  });
}

void _expectTransportProbe(String environmentName) {
  final configured = Platform.environment[environmentName];
  if (configured == null || configured.isEmpty) {
    markTestSkipped('$environmentName is required for governed transport execution');
    return;
  }
  final client = NativeClient.load(path.normalize(path.absolute(configured)));
  try {
    expect(
      client.describe,
      throwsA(
        isA<NativeAdapterException>().having(
          (error) => error.kind,
          'kind',
          NativeErrorKind.transport,
        ),
      ),
    );
  } finally {
    client.close();
  }
}

void _writeEvidence(Map<String, Object?> observations) {
  final root = Platform.environment['STRLING_GDS_EVIDENCE_DIR'];
  if (root == null || root.isEmpty) return;
  final directory = Directory(path.join(root, 'dart'))..createSync(recursive: true);
  for (final entry in observations.entries) {
    File(path.join(directory.path, '${entry.key}.json'))
        .writeAsStringSync(jsonEncode(entry.value));
  }
}
